File: javascript/src/change.js
// Welcome to qq group: 1030115250

// Welcome to qq group: 1030115250
const {decryptData,encryptJSON} = require('./aesnet.js');
const {hostaddr, netdata,keyBuffer} = require('./host.js');
const crypto = require('crypto');
const {sendEncryptRequest} = require('./login.js');

  async function changepassword(username, newpassword,question,answer) {
    try {
        const init = await sendEncryptRequest();
        const decrypted = decryptData(init.data, keyBuffer, init.iv, init.authTag);
        netdata.aeskey = Buffer.from(decrypted.data, 'base64'); // 明确转换为 Buffer
        const newpasswordHash = crypto.createHash('sha512').update(newpassword).digest('hex');
        const questionHash = crypto.createHash('sha512').update(question).digest('hex');
        const answerHash = crypto.createHash('sha512').update(answer).digest('hex');
        const data = { 
            name:username, 
            newpassword:newpasswordHash,
            question:questionHash,
            answer:answerHash,
            timestamp:Date.now(),
        };
        const encryptedData = encryptJSON(data, netdata.aeskey);
        const response = await fetch(`${hostaddr}/api/users/change`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body:JSON.stringify(encryptedData)
          });

        if (!response.ok) {
            const errorData = await response.json();
            throw new Error(`请求失败: ${errorData.error} (原因: ${errorData.reason})`);
        }    

        const responseData = await response.json();
        const deData = decryptData(responseData.data, netdata.aeskey, responseData.iv, responseData.tag); 
        if (deData?.newkey){
            netdata.aeskey = Buffer.from(deData.newkey, 'base64'); // 明确转换为 Buffer
        }
        console.log('修改密码成功:', deData);
        return deData.success;
    }
    catch (error) {
      console.error('修改密码失败:', error); 
      throw error;
    }
  }

  // changepassword('testAdmin', 'MySecret1234','Safe456');

  module.exports = {changepassword};

File: Python/src/change.py
import base64
import hashlib
import time
import requests
from .host import hostaddr, netdata, keyBuffer
from .aesnet import decrypt_data, encrypt_json
from .login import send_encrypt_request

def changepassword(username, newpassword, question, answer):
    try:
        # 发送初始化请求获取加密参数
        init = send_encrypt_request()
        
        # 解密初始化响应获取AES密钥
        decrypted = decrypt_data(
            ciphertext=init['data'],
            current_key=keyBuffer,
            iv=init['iv'],
            tag=init['authTag']
        )
        netdata['aeskey'] = base64.b64decode(decrypted['data'])  # 解码base64密钥
        
        # 生成SHA-512哈希
        newpassword_hash = hashlib.sha512(newpassword.encode()).hexdigest()
        question_hash = hashlib.sha512(question.encode()).hexdigest()
        answer_hash = hashlib.sha512(answer.encode()).hexdigest()
        
        # 构造请求数据
        payload = {
            'name': username,
            'newpassword': newpassword_hash,
            'question': question_hash,
            'answer': answer_hash,
            'timestamp': int(time.time() * 1000)  # 毫秒级时间戳
        }
        
        # AES-GCM加密数据
        encrypted_data = encrypt_json(payload, netdata['aeskey'])
        
        # 发送修改密码请求
        response = requests.post(
            f"{hostaddr}/api/v1/users/change",
            headers={'Content-Type': 'application/json'},
            json=encrypted_data
        )
        response.raise_for_status()  # 检查HTTP错误
        
        # 解密响应数据
        response_data = response.json()
        de_data = decrypt_data(
            ciphertext=response_data['data'],
            current_key=netdata['aeskey'],
            iv=response_data['iv'],
            tag=response_data['tag']
        )
        
        # 轮换新密钥（如果有）
        if 'newkey' in de_data:
            netdata['aeskey'] = base64.b64decode(de_data['newkey'])
        
        print('密码修改成功:', de_data)
        return de_data.get('success', False)
    
    except requests.exceptions.RequestException as e:
        print(f'网络请求异常: {e}')
        raise
    except Exception as e:
        print(f'操作失败: {e}')
        raise